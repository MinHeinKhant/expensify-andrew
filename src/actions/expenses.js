import uuid from 'uuid';
import database from '../firebase/firebase';

// w/o firebase
// component calls function generator
// action generator returns object
// component dispatch object
// redux store changes

// with firebase
// components calls action generator
// action generator returns function
// component dispatches function (using middleware)
// function run (has the ability to dispatch other actions and do whatever it wants)

// ADD_EXPENSE
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

// with firebase
export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = {
      description,
      note,
      amount,
      createdAt
    };

    const uid = getState().auth.uid;

    return database
      .ref(`users/${uid}/expenses`)
      .push(expense)
      .then(ref => {
        dispatch(addExpense({ id: ref.key, ...expense }));
      });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    database
      .ref(`users/${uid}/expenses/${id}`)
      .remove()
      .then(() => {
        console.log('removed from database');
        dispatch(removeExpense({ id }));
      });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    database
      .ref(`users/${uid}/expenses/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editExpense(id, updates));
      });
  };
};

// SET_EXPENSES
export const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    return database
      .ref(`users/${uid}/expenses`)
      .once('value')
      .then(snapshot => {
        const expenses = [];
        snapshot.forEach(childSnapshot => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });

        dispatch(setExpenses(expenses));
      });
  };
};
