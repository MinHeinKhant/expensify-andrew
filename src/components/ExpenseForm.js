import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
    this.onDescriptionChange = this.onDescriptionChange.bind(this);
    this.onNoteChange = this.onNoteChange.bind(this);
    this.onAmountChange = this.onAmountChange.bind(this);
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({
      description
    }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({
        amount
      }));
    }
  };

  onNoteChange = e => {
    const note = e.target.value;
    // Use it when you use note: e.target.vlaue
    // e.persist();
    this.setState(() => ({
      note
    }));
  };

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({
        createdAt
      }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({
      calendarFocused: focused
    }));
  };

  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      // Set error state equals to 'Please provide description and amount'
      this.setState(() => ({
        error: 'Please provide description and amount'
      }));
    } else {
      // Clear the error
      this.setState(() => ({
        error: ''
      }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          className="text-input"
          value={this.state.description}
          type="text"
          placeholder="Description"
          autoFocus
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          className="text-input"
          value={this.state.amount}
          onChange={this.onAmountChange}
          placeholder="Amount"
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          className="text-area"
          value={this.state.note}
          onChange={this.onNoteChange}
          placeholder="Enter your notes"
        />
        <div>
          <button className="button">Add Expense</button>
        </div>
      </form>
    );
  }
}
