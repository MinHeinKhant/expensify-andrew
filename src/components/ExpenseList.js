import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpense from '../selectors/expenses';

export const ExpenseList = props => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expense</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-body">
      {props.expenses.length === 0 ? (
        <div>
          <span className="list-item--message">No Expense</span>
        </div>
      ) : (
        props.expenses.map(expense => {
          return <ExpenseListItem key={expense.id} {...expense} />;
        })
      )}
    </div>
  </div>
);

const mapStateToProps = state => {
  return {
    expenses: selectExpense(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
