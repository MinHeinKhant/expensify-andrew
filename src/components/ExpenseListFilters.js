import React from 'react';
import { connect } from 'react-redux';
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate
} from '../actions/filters';
import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css';
import { DateRangePicker } from 'react-dates';

class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    // this.setState(() => ({
    //   startDate: this.props.dispatch(setStartDate(startDate)),
    //   endDate: this.props.dispatch(setEndDate(endDate))
    // }));

    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  };
  onFocusChange = calendarFoucused => {
    this.setState(() => ({ calendarFoucused }));
  };
  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              placeholder="Search Expenses"
              className="text-input"
              type="text"
              value={this.props.filters.text}
              onChange={e => {
                this.props.dispatch(setTextFilter(e.target.value));
              }}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={e => {
                if (e.target.value === 'date') {
                  this.props.dispatch(sortByDate());
                } else if (e.target.value === 'amount') {
                  this.props.dispatch(sortByAmount());
                }
              }}
            >
              <option value="amount">Amount</option>
              <option value="date">Date</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDateId="startDate"
              endDateId="endDate"
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.focusedInput}
              onFocusChange={focusedInput => {
                this.setState({ focusedInput });
              }}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    filters: state.filters
  };
};

export default connect(mapStateToProps)(ExpenseListFilters);
