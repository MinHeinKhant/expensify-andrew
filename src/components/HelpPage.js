import React from 'react';

const HelpPage = () => <div>This is my help page</div>;

export default HelpPage;
