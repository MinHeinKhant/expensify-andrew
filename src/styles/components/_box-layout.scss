.box-layout {
  background: url('/images/bg.jpg');
  background-size: cover;
  height: 100vh;
  width: 100vw;
  display: flex;
  justify-content: center;
  align-items: center;
}

.box-layout__box {
  background: fade-out(white, 0.13);
  border-radius: 3px;
  text-align: center;
  width: 25rem;
  padding: $l-size $m-size;
}

.box-layout__title {
  margin: 0 0 $m-size 0;
  line-height: 1;
}
