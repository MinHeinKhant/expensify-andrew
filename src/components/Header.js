import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = ({ startLogout }) => (
  <header>
    <h1>Expensify</h1>

    <NavLink activeClassName="is-active" to="/" exact={true}>
      Dashboard
    </NavLink>
    <NavLink activeClassName="is-active" to="/create">
      Create Expense
    </NavLink>
    {/* <NavLink activeClassName="is-active" to="/help">
      Help
</NavLink> */}
    <button onClick={startLogout}>Logout</button>
  </header>
);

const mapDispatchToProps = dispatch => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(
  undefined,
  mapDispatchToProps
)(Header);
