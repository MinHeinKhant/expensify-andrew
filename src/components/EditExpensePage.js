import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense } from '../actions/expenses';
import { startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = expense => {
    // Dispatch the action to edit the expense
    this.props.dispatch(startEditExpense(this.props.expense.id, expense));

    // Redirect to the dashboardpage
    this.props.history.push('/');
  };

  onRemove = () => {
    console.log('removed');
    this.props.dispatch(startRemoveExpense({ id: this.props.expense.id }));
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />

        <button onClick={this.onRemove}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

// const mapDispatchToProps = (dispatch, props) => ({
//   editExpense: (id, expense) => dispatch(editExpense(id, expense)),
//   startRemoveExpense: data => dispatch(startRemoveExpense(data))
// });

export default connect(mapStateToProps)(EditExpensePage);
