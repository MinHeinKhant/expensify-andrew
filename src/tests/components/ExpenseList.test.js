import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

test('should render ExpenseList with expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});

test('should render ExpenseList without expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});
