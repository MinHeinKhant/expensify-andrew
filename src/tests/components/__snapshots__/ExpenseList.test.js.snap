// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="Gum"
    id="0"
    key="0"
    note=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={-345600000}
    description="Rent"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={4500}
    createdAt={345600000}
    description="Credit Card"
    id="2"
    key="2"
    note=""
  />
</div>
`;

exports[`should render ExpenseList without expenses 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;
